src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -0,0 +1,317 @@
+         *
+         * @param pointer Pointer in memory to the index
+         * @param size Size this index consumes in kilobytes
+         * @param knnEngine KNNEngine associated with the index allocation
+         * @param indexPath File path to index
+         * @param osIndexName Name of OpenSearch index this index is associated with
+         * @param watcherHandle Handle for watching index file
+         */
+        public IndexAllocation(long pointer, long size, KNNEngine knnEngine, String indexPath, String osIndexName,
+                               WatcherHandle<FileWatcher> watcherHandle) {
+            this.closed = false;
+            this.knnEngine = knnEngine;
+            this.indexPath = indexPath;
+            this.osIndexName = osIndexName;
+            this.pointer = pointer;
+            this.readWriteLock = new ReentrantReadWriteLock();
+            this.size = size;
+            this.watcherHandle = watcherHandle;
+        }
+
+        @Override
+        public void close() {
+            // Lock acquisition should be done by caller
+            if (this.closed) {
+                return;
+            }
+
+            try {
+                watcherHandle.stop();
+
+                // Pointer is sometimes initialized to 0. If this is ever the case, freeing will surely fail.
+                if (pointer != 0) {
+                    JNIService.free(pointer, knnEngine.getName());
+                }
+            } finally {
+                this.closed = true;
+            }
+        }
+
+        @Override
+        public boolean isClosed() {
+            return closed;
+        }
+
+        @Override
+        public long getPointer() {
+            return pointer;
+        }
+
+        @Override
+        public void readLock() {
+            readWriteLock.readLock().lock();
+        }
+
+        @Override
+        public void writeLock() {
+            readWriteLock.writeLock().lock();
+        }
+
+        @Override
+        public void readUnlock() {
+            readWriteLock.readLock().unlock();
+        }
+
+        @Override
+        public void writeUnlock() {
+            readWriteLock.writeLock().unlock();
+        }
+
+        @Override
+        public long getSizeInKb() {
+            return size;
+        }
+
+        /**
+         * Getter for k-NN Engine associated with this index allocation.
+         *
+         * @return KNNEngine associated with index allocation
+         */
+        public KNNEngine getKnnEngine() {
+            return knnEngine;
+        }
+
+        /**
+         * Getter for the path to the file from which the index was loaded.
+         *
+         * @return indexPath to index
+         */
+        public String getIndexPath() {
+            return indexPath;
+        }
+
+        /**
+         * Getter for the OpenSearch index associated with the native index.
+         *
+         * @return OpenSearch index name
+         */
+        public String getOsIndexName() {
```

Comment:
os is not preferred , we should use openserach.

Reply:
Will refactor